### PR TITLE
PF-241: Use Test Runner library to write tests using client library.

### PR DESCRIPTION
**Pull request by marikomedlock**
This PR contains the first pass at using the Test Runner library to write client tests for Buffer Service.

- Added a buffer-clienttests Gradle project.
- Included a server specification for a developer namespace in the terra-integration cluster.
- Wrote one test that hits the unauthenticated endpoint, and one that hits the unauthenticated endpoint.

Also changed the git clone commands in the deploy script to use the .git version of the repository address, instead of https.

### CODE REVIEW

File: buffer-clienttests/build.gradle [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+sourceCompatibility = 1.8
+
+spotless {
+    java {
+        googleJavaFormat()
+    }
+}
+compileJava.dependsOn tasks.spotlessApply
+
+spotbugs {
```
Comment (yonghaoy) [inline buffer-clienttests/build.gradle:69]: I think recently Issac standerlize WM/Janitor/RBS build.gradle file. And they don't have those spotbugs stuff. Do we need those? Shall we follow the same pattern like RBS's build.gradle or we should add this into the standard?

**Review by yonghaoy — CHANGES_REQUESTED** [review]
Thanks for doing this! 
Overrall I have some questions about how we can use it?  For now, this seems to have some to many SA json files and not well managed. I don't think it will work for me or other developers. I think we need a better piping secret & config piping scenarios here. 
Happy to chat if you have time after standup. Woud be great if @gjuggler can also help on those configs management stuff. 

Thanks!

File: local-dev/setup_gke_deploy.sh [hunk 2]
```diff
@@ -15,8 +15,8 @@ TERRA_HELMFILE_BRANCH=${3:-master}
 # Clone Helm chart and helmfile repos
 rm -rf terra-helm
 rm -rf terra-helmfile
-git clone -b "$TERRA_HELM_BRANCH" --single-branch https://github.com/broadinstitute/terra-helm
-git clone -b "$TERRA_HELMFILE_BRANCH" --single-branch https://github.com/broadinstitute/terra-helmfile
+git clone -b "$TERRA_HELM_BRANCH" --single-branch git@github.com:broadinstitute/terra-helm.git
```
[the remark this answers is not in the document]
Comment (gjuggler) [reply local-dev/setup_gke_deploy.sh:18]: cloning from HTTPS URLs for private GitHub repos requires interactively typing username and password on the command-line. That feels like an awkward side-effect of a setup script, and it's also painful for folks who use a password for GitHub that isn't in muscle memory.

What problems did you run into getting set up with SSH for GitHub? AFAIK it's a very common setup — it may be worth working through solutions for those issues.

(Obviously this is separate thread from this PR, we can follow-up after standup or in slack)

[the remark this answers is not in the document]
Comment (yonghaoy) [reply local-dev/setup_gke_deploy.sh:18]: Somehow I don't need to type my username and password when using https urls. But I got some authentication issue when using ssh. I think it's same for Will recently:
https://github.com/DataBiosphere/terra-resource-janitor/pull/62#issuecomment-729756525
I can try to setup my ssh again.. There might be some steps we missed.

[the remark this answers is not in the document]
Comment (marikomedlock) [reply local-dev/setup_gke_deploy.sh:18]: Okay, I'll put it back to HTTPS until we have consensus on the approach in Terra code.

File: buffer-clienttests/src/main/resources/testusers/neville.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "neville",
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/resources/testusers/neville.json:2]: Per our conversation last week, I'll remove these users and the associated SA since you're not planning to use domain-wide delegated credentials for Buffer Service.

For the other questions:
- These test users are not Data Repo-specific. All services that end users can talk to directly will probably want to do some testing with user credentials rather than service credentials. Currently in Data Repo (and my understanding also other existing Terra services), we do that with domain-wide delegated user credentials.
- Domain-wide delegated user credentials require a SA that can get credentials on behalf of a user. We cannot create this in Terraform because domain-wide delegation is a GSuite thing, not a GCP thing, and requires some manual steps to setup. For this reason, I think we should just use one SA that can get domain-wide delegated credentials for all services. Data Repo already has one that we use for this -- it isn't used anywhere else.

File: buffer-clienttests/src/main/resources/serviceaccounts/jade-k8-sa.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "jade-k8-sa@broad-jade-dev.iam.gserviceaccount.com",
+  "jsonKeyFilename": "jade-dev-account.json",
+  "jsonKeyDirectoryPath": "/tmp"
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/resources/serviceaccounts/jade-k8-sa.json:4]: This SA is to support domain-wide delegation. Per our conversation last week, I'll remove this since you're not planning to use domain-wide delegation to test Buffer Service.

File: buffer-clienttests/src/main/java/scripts/utils/BufferServiceUtils.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    AccessToken accessToken = AuthenticationUtils.getAccessToken(serviceAccountCredential);
+    logger.debug(
+        "Generated access token for buffer service client SA: {}",
+        server.bufferClientServiceAccount.name);
+
+    // build the client object
+    ApiClient apiClient = new ApiClient();
+    apiClient.setBasePath(server.bufferUri);
+    apiClient.setAccessToken(accessToken.getTokenValue());
+    // apiClient.setDebugging(true); // useful for debugging problems header sent/received
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/java/scripts/utils/BufferServiceUtils.java:44]: Removed, it was just for debugging.

File: buffer-clienttests/src/main/resources/configs/BasicAuthenticated.json [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  "serverSpecificationFile": "yyu-dev.json",
+  "billingAccount": "00708C-45D19D-27AAFA",
+  "kubernetes": {},
+  "application": {},
+  "testScripts": [
+    {
+      "name": "GetPoolInfo",
+      "numberOfUserJourneyThreadsToRun": 10,
+      "userJourneyThreadPoolSize": 2,
+      "expectedTimeForEach": 5,
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/resources/configs/BasicAuthenticated.json:13]: Yes, we could always specify things in seconds or whatever we want for the smallest time unit. Since some of the user journeys in Data Repo could take a long time (~hours), it made the test specifications more readable to use larger units.

File: buffer-clienttests/src/main/java/scripts/testscripts/ServiceStatus.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+public class ServiceStatus extends TestScript {
+  private static final Logger logger = LoggerFactory.getLogger(ServiceStatus.class);
+
+  /** Public constructor so that this class can be instantiated via reflection. */
+  public ServiceStatus() {
+    super();
+  }
+
+  public void userJourney(TestUserSpecification testUser) throws Exception {
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/java/scripts/testscripts/ServiceStatus.java:20]: I don't follow your comment. userJourney already is a method of TestScript (the base class that ServiceStatus extends) that must be overridden. Can you rephrase?

File: buffer-clienttests/build.gradle (hunk 1 above)
Comment (marikomedlock) [reply buffer-clienttests/build.gradle:71]: I used spotbugs because it is in the MC Terra Tech Stack [document](https://docs.google.com/document/d/1JkTrtaci7EI0TnuR-68zYgTx_mRCNu-u2eV9XhexWTI/edit#heading=h.dndcg044j2p). What are you using for static analysis instead?

File: buffer-clienttests/settings.gradle [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        resolutionStrategy {
+            eachPlugin {
+                if (requested.id.namespace == "bio.terra") {
+                    if (requested.id.id == "bio.terra.test-runner-plugin" || requested.id.id == "bio.terra.test-runner-echo-plugin") {
+                        useModule("bio.terra:terra-test-runner:" + requested.version)
+                    }
+                }
+            }
+        }
+        // To use a local copy of the Test Runner library, uncomment the flatDir block and comment out the maven block
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/settings.gradle:12]: Removed debugging code.

File: buffer-clienttests/src/main/java/scripts/utils/BufferServiceUtils.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      throw new IllegalArgumentException("Buffer Service URI cannot be empty");
+    }
+    if (server.bufferClientServiceAccount == null) {
+      throw new IllegalArgumentException("Buffer Service client service account is required");
+    }
+
+    // refresh the client service account token
+    GoogleCredentials serviceAccountCredential =
+        AuthenticationUtils.getServiceAccountCredential(server.bufferClientServiceAccount, true);
+    AccessToken accessToken = AuthenticationUtils.getAccessToken(serviceAccountCredential);
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/java/scripts/utils/BufferServiceUtils.java:35]: Yes, scopes are set on the SA credential to the (user login) scopes "openid", "email", "profile".

File: buffer-clienttests/src/main/java/scripts/utils/BufferServiceUtils.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  private BufferServiceUtils() {}
+
+  /**
+   * Build the Buffer Service API client object for the given server specification.
+   *
+   * @param server the server we are testing against
+   * @return the API client object
+   */
+  public static ApiClient getClient(ServerSpecification server) throws IOException {
+    if (server.bufferUri == null || server.bufferUri.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/java/scripts/utils/BufferServiceUtils.java:24]: I don't see Guava listed in the MC Terra tech stack [document](https://docs.google.com/document/d/1JkTrtaci7EI0TnuR-68zYgTx_mRCNu-u2eV9XhexWTI/edit#).

But it seems services can make their own decisions about which libraries to include, so I added this dependency to the buffer-clienttests project.

File: buffer-clienttests/settings.gradle [hunk 11]
```diff
@@ -0,0 +1,23 @@
+pluginManagement {
+    repositories {
+        resolutionStrategy {
+            eachPlugin {
+                if (requested.id.namespace == "bio.terra") {
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/settings.gradle:5]: This logic tells Gradle where to find the JAR that contains the TestRunner Gradle plugin. The purpose of the plugin is so that each project that contains tests doesn't have to copy/paste the same Gradle tasks.

File: buffer-clienttests/src/main/java/scripts/utils/BufferServiceUtils.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import bio.terra.buffer.client.ApiClient;
+import bio.terra.testrunner.common.utils.AuthenticationUtils;
+import bio.terra.testrunner.runner.config.ServerSpecification;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.GoogleCredentials;
+import java.io.IOException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BufferServiceUtils {
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/java/scripts/utils/BufferServiceUtils.java:12]: Added.

File: buffer-clienttests/src/main/resources/configs/BasicUnauthenticated.json [hunk 13]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "BasicUnauthenticated",
+  "description": "Check the service status repeatedly, concurrently. No authentication required.",
+  "serverSpecificationFile": "yyu-dev.json",
+  "billingAccount": "00708C-45D19D-27AAFA",
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/resources/configs/BasicUnauthenticated.json:5]: Data Repo clients need to specify a billing account to use when creating resources, so this is just a test configuration property. It's not required though, so I'll just remove it since Buffer Service isn't using it now.

File: buffer-clienttests/src/main/resources/configs/BasicUnauthenticated.json [hunk 14]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "BasicUnauthenticated",
+  "description": "Check the service status repeatedly, concurrently. No authentication required.",
+  "serverSpecificationFile": "yyu-dev.json",
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/resources/configs/BasicUnauthenticated.json:4]: Yes, the server specification can be [overridden](https://github.com/DataBiosphere/terra-test-runner#override-the-server-from-the-command-line) by an environment variable.

File: buffer-clienttests/src/main/resources/serviceaccounts/mmedlock-buffer-client-sa.json [hunk 15]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "buffer-client-mmedlock@terra-kernel-k8s.iam.gserviceaccount.com",
+  "jsonKeyFilename": "sa-account.json",
+  "jsonKeyDirectoryPath": "../src/test/resources/rendered"
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/resources/serviceaccounts/mmedlock-buffer-client-sa.json:4]: Okay so do I need a separate render-configs script for testing that pulls the RBS client SA account from Vault?

File: buffer-clienttests/src/main/resources/servers/mmedlock-dev.json [hunk 16]
```diff
@@ -0,0 +1,17 @@
+{
```
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/resources/servers/mmedlock-dev.json:1]: An instance of the ServerSpecification class (= JSON files in the resources/servers directory) represents a single Terra environment or deployment. So we should have a single JSON file for each environment that we want to run tests against. I was thinking this will include at least one developer namespace in terra-integration cluster where you can debug, and one for each environment where performance/resilience tests need to be run periodically. In Data Repo, this includes perf, alpha, and staging.

**Comment by marikomedlock** [comment]
Thanks for the review. I marked the conversations resolved where I didn't have a question - please reopen with any follow-ups.

Agreed about needing a better way to manage config/secrets piping. However, I don't think that should block development of client tests. In the long-term, I think we should be able to point to a Terra environment/deployment with a single keyword (e.g. terra-integration-mmedlock, rbs-perf, preview-env-123) and then derive all the other information relevant to clients from that. Relevant information for clients includes: service urls, client SAs (or maybe Vault paths to their credentials), any additional information about connections between the services (e.g. SAM resource id that corresponds to the particular Data Repo instance). One way to derive all this information is to look it up in a Kubernetes config map.

Since we don't have this config/secrets management solution yet, the Test Runner library uses a manual specification of a Terra environment/deployment. The intention is that we can replace this manual specification with a dynamic lookup of Terra environment properties (e.g. service URLs) once the Kubernetes config map (or other) solution is available.

This is a summary of my thoughts on this. But maybe this topic is better to discuss in a meeting. @yonghaoy What do you think?

File: buffer-clienttests/src/main/resources/serviceaccounts/mmedlock-buffer-client-sa.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (yonghaoy) [reply buffer-clienttests/src/main/resources/serviceaccounts/mmedlock-buffer-client-sa.json:4]: Yes, I think we need a script like this: https://github.com/DataBiosphere/terra-resource-buffer/blob/master/local-dev/render-config.sh
And I suggest to create a new script for the test runner SAs in its own package.

File: buffer-clienttests/src/main/java/scripts/testscripts/ServiceStatus.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (yonghaoy) [reply buffer-clienttests/src/main/java/scripts/testscripts/ServiceStatus.java:20]: Ah, I didn't realize userJourney already a method of TestScript. If that is the case, I think you miss `@override` annotation here.
Should be 
```
@Override
public void userJourney(TestUserSpecification testUser) throws Exception {
```

File: buffer-clienttests/build.gradle (hunk 1 above)
Comment (yonghaoy) [reply buffer-clienttests/build.gradle:73]: Ah, I see. Then I think RBS & WM are missing this thing

**Comment by yonghaoy** [comment]
Thanks! @marikomedlock Overrall the test runner looks good to me. But I am little bit confused about how to actually make this thing useful. This including:

1. How /when / where to trigger this test runner. (in release process? nightly run? or something else)? 
2. What envs we need for this and what additional the CI/CD work we need?
3. I think we won't have preview env ready recently but RBS needs to go to production soon, how can we use this without preview environment? 

Thanks

File: buffer-clienttests/src/main/resources/serviceaccounts/mmedlock-buffer-client-sa.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/resources/serviceaccounts/mmedlock-buffer-client-sa.json]: Got it. I added one, PTAL.

File: buffer-clienttests/src/main/java/scripts/testscripts/ServiceStatus.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (marikomedlock) [reply buffer-clienttests/src/main/java/scripts/testscripts/ServiceStatus.java:21]: Ah I understand. Yes good point. Added the annotations.

**Comment by marikomedlock** [comment]
Thanks for the helpful comments and pointer to the render-config example.

Test Runner should run in:
- GH actions. I don't know what the frequency should be -- I think that is independent of Test Runner, and probably we should have some conversation with Greg or QA about the recommended approach.
- Local machine. This is for writing and debugging tests, and for running one-time profiling experiments (e.g. to determine appropriate config value)

Agreed about not waiting for preview environments to be ready. I think we should run Test Runner tests against environments/deployments that are already setup manually (e.g. developer namespace, perf/alpha/staging/dev). This is what Data Repo currently does.

With this branch, you should be able to write a performance test and run it against an existing environment. This was the first goal. There are some more changes coming:
- Ivan will probably write some example GH actions, based on existing Data Repo examples. Once those are ready, we can put an example in this branch.
- We will make some changes to Test Runner library to support resilience tests. Once that is ready, we can also put an example in this branch.

So I'll plan to leave this branch here, not merged, and re-request your review when the above changes are ready. In the meantime, you can try to write performance tests if you'd like, or wait until all features are available.

**Review by yonghaoy — APPROVED** [review]
You can go ahead and merge it, and I will have a follow up PR to setup it in github action.

Thanks!
